File: app/src/main/java/com/appchamp/wordchunks/ui/game/GameActivity.java
package com.appchamp.wordchunks.ui.game;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.view.View;

import com.appchamp.wordchunks.R;
import com.appchamp.wordchunks.data.LevelsRealmHelper;
import com.appchamp.wordchunks.data.PacksRealmHelper;
import com.appchamp.wordchunks.models.realm.Level;
import com.appchamp.wordchunks.models.realm.Pack;
import com.appchamp.wordchunks.ui.game.fragments.GameFinishedFrag;
import com.appchamp.wordchunks.ui.game.fragments.GameFrag;
import com.appchamp.wordchunks.ui.game.fragments.LevelSolvedBeforeFrag;
import com.appchamp.wordchunks.ui.game.fragments.LevelSolvedFrag;
import com.appchamp.wordchunks.ui.game.listeners.OnBackToLevelsListener;
import com.appchamp.wordchunks.ui.game.listeners.OnLevelSolvedListener;
import com.appchamp.wordchunks.ui.game.listeners.OnNextLevelListener;
import com.appchamp.wordchunks.ui.levels.LevelsActivity;
import com.appchamp.wordchunks.ui.tutorial.TutorialActivity;
import com.appchamp.wordchunks.util.ActivityUtils;
import com.appchamp.wordchunks.util.AnimUtils;

import io.realm.Realm;
import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;

import static com.appchamp.wordchunks.util.Constants.EXTRA_LEVEL_ID;
import static com.appchamp.wordchunks.util.Constants.EXTRA_PACK_ID;
import static com.appchamp.wordchunks.util.Constants.LEVEL_STATE_CURRENT;
import static com.appchamp.wordchunks.util.Constants.LEVEL_STATE_LOCKED;
import static com.appchamp.wordchunks.util.Constants.LEVEL_STATE_SOLVED;
import static com.appchamp.wordchunks.util.Constants.PACK_STATE_CURRENT;
import static com.appchamp.wordchunks.util.Constants.PACK_STATE_LOCKED;
import static com.appchamp.wordchunks.util.Constants.PACK_STATE_SOLVED;
import static com.appchamp.wordchunks.util.Constants.PREFS_HOW_TO_PLAY;
import static com.appchamp.wordchunks.util.Constants.WORD_CHUNKS_PREFS;


public class GameActivity extends AppCompatActivity implements OnLevelSolvedListener,
        OnNextLevelListener, OnBackToLevelsListener {

    private String levelId;
    private Level nextLevel;
    private Realm realm;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.act_game);
        realm = Realm.getDefaultInstance();

        SharedPreferences sp = getSharedPreferences(WORD_CHUNKS_PREFS, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();

        if (sp.getBoolean(PREFS_HOW_TO_PLAY, true)) {
            // show how to play
            showTutorial();
            editor.putBoolean(PREFS_HOW_TO_PLAY, false);
            editor.apply();
        }
        addGameFragment();
    }


    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        startLevelsActivity();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        realm.close(); // Remember to close Realm when done.
    }

    public void onBackArrowClick(View v) {
        super.onBackPressed();
        AnimUtils.startAnimationFadeIn(GameActivity.this, v);
        startLevelsActivity();
    }

    private void showTutorial() {
        Intent intent = new Intent(this, TutorialActivity.class);
        startActivity(intent);
    }

    private void addGameFragment() {
        // Getting level id via Intents
        levelId = getIntent().getStringExtra(EXTRA_LEVEL_ID);

        ActivityUtils.addFragment(
                getSupportFragmentManager(),
                GameFrag.newInstance(),
                R.id.flActMain,
                EXTRA_LEVEL_ID,
                levelId);
    }

    /**
     * Back navigation. Navigates from GameActivity to LevelsActivity passing Pack id in the Intent.
     */
    private void startLevelsActivity() {
        Intent intent = new Intent(GameActivity.this, LevelsActivity.class);

        // Passing levelId via Intent.
        String packId = LevelsRealmHelper.findLevelById(realm, levelId).getPackId();
        intent.putExtra(EXTRA_PACK_ID, packId);

        startActivity(intent);
        overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right);
    }

    /**
     * This callback method called from the GameFrag immediately when the level was solved.
     */
    @Override
    public void onLevelSolved() {
        realm.executeTransaction(bgRealm -> {

            Level level = LevelsRealmHelper.findLevelById(bgRealm, levelId);
            int packColor = Color.parseColor(
                    PacksRealmHelper.findFirstPackById(bgRealm, level.getPackId())
                            .getColor());

            if (isLevelSolvedBefore(level)) {
                // Level WAS solved before
                showLevelSolvedBeforeFragment(packColor, level.getFact());

            } else {
                // Level was NOT solved before
                // Change current level state to "solved", current level state is "current"
                level.setState(LEVEL_STATE_SOLVED);

                // Unlocking the next level to play in.

                // Getting the next level to play in.
                nextLevel = LevelsRealmHelper.findFirstLevelByState(bgRealm, LEVEL_STATE_LOCKED);
//                Logger.d("Next Level id  = " + nextLevel.getId());

                String nextLevelClue = "";
                // If the next locked level exists
                if (nextLevel != null) {
                    // Unlock next level
                    nextLevel.setState(LEVEL_STATE_CURRENT);
                    nextLevelClue = nextLevel.getClue();
                }

                showLevelSolvedFragment(
                        packColor,
                        nextLevelClue,
                        level.getFact(),
                        isPackSolved(bgRealm, level.getPackId()));
            }
        });
    }

    private boolean isLevelSolvedBefore(Level level) {
        // Getting the state of the current level, if "current": return false, else true.
        return level.getState() != LEVEL_STATE_CURRENT;
    }

    private long isPackSolved(Realm bgRealm, String packId) {
        // Count the number of current levels in pack, if 0 then we solved the whole pack

        if (LevelsRealmHelper
                .countLevelsByPackIdAndState(bgRealm, packId, LEVEL_STATE_CURRENT) == 0) {

            PacksRealmHelper.findFirstPackById(bgRealm, packId).setState(PACK_STATE_SOLVED);
            // Find the next locked pack and set it as "current"
            Pack nextLockedPack = PacksRealmHelper.findFirstPackByState(bgRealm, PACK_STATE_LOCKED);
            if (nextLockedPack != null) {
                nextLockedPack.setState(PACK_STATE_CURRENT);
            }
            return 0;
        } else {
            long numberOfSolvedLevels = LevelsRealmHelper
                    .countLevelsByPackIdAndState(bgRealm, packId, LEVEL_STATE_SOLVED);
            long numberOfLevels = LevelsRealmHelper
                    .countLevelsByPackId(bgRealm, packId);
            return numberOfLevels - numberOfSolvedLevels;
        }
    }

    private void showLevelSolvedBeforeFragment(int color, String fact) {
        Bundle args = new Bundle();
        Fragment fragment = LevelSolvedBeforeFrag.newInstance();
        args.putInt("color", color);
        args.putString("fact", fact);
        fragment.setArguments(args);
        ActivityUtils.replaceFragment(
                getSupportFragmentManager(),
                fragment,
                R.id.flActMain);
    }

    private void showGameFinishedFragment() {
        ActivityUtils.replaceFragment(
                getSupportFragmentManager(),
                GameFinishedFrag.newInstance(),
                R.id.flActMain);
    }

    private void showLevelSolvedFragment(int color, String clue, String fact, long left) {
        Bundle args = new Bundle();
        Fragment fragment = LevelSolvedFrag.newInstance();
        args.putInt("color", color);
        args.putString("clue", clue);
        args.putString("fact", fact);
        args.putLong("left", left);
        fragment.setArguments(args);
        ActivityUtils.replaceFragment(
                getSupportFragmentManager(),
                fragment,
                R.id.flActMain);
    }

    @Override
    public void onNextLevelSelected() {
        if (nextLevel != null) {
            levelId = nextLevel.getId();
            ActivityUtils.replaceFragment(
                    getSupportFragmentManager(),
                    GameFrag.newInstance(),
                    R.id.flActMain,
                    EXTRA_LEVEL_ID,
                    levelId);
        } else {
            showGameFinishedFragment();
        }
    }

    @Override
    public void onBackToLevelsSelected() {
        super.onBackPressed();
        startLevelsActivity();
    }
}

File: app/src/main/java/com/appchamp/wordchunks/ui/game/GameActivity.kt
/*
 * Copyright 2017 Julia Kozhukhovskaya
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.appchamp.wordchunks.ui.game

import android.arch.lifecycle.Observer
import android.content.SharedPreferences
import android.graphics.Color
import android.os.Bundle
import android.view.Gravity
import android.widget.Toast
import com.appchamp.wordchunks.R
import com.appchamp.wordchunks.models.realm.FINISHED
import com.appchamp.wordchunks.models.realm.IN_PROGRESS
import com.appchamp.wordchunks.models.realm.Level
import com.appchamp.wordchunks.ui.customviews.LevelSolvedDialog
import com.appchamp.wordchunks.ui.customviews.StoreDialog
import com.appchamp.wordchunks.ui.levels.LevelsActivity
import com.appchamp.wordchunks.util.ActivityUtils
import com.appchamp.wordchunks.util.Constants.EXTRA_LEVEL_ID
import com.appchamp.wordchunks.util.Constants.EXTRA_PACK_ID
import com.appchamp.wordchunks.util.Constants.PREFS_NAME
import com.appchamp.wordchunks.util.Constants.PREFS_TUTORIAL
import com.appchamp.wordchunks.util.Constants.USER_DAILY_LEVEL_SOLVED_REWARD
import com.appchamp.wordchunks.util.Constants.USER_LEVEL_SOLVED_REWARD
import kotlinx.android.synthetic.main.frag_game.*
import kotlinx.android.synthetic.main.titlebar.*
import me.toptas.fancyshowcase.FancyShowCaseQueue
import me.toptas.fancyshowcase.FancyShowCaseView
import me.toptas.fancyshowcase.FocusShape
import org.jetbrains.anko.clearTop
import org.jetbrains.anko.intentFor


class GameActivity : BaseGameActivity(), LevelSolvedDialog.LevelSolvedDialogListener,
        StoreDialog.StoreDialogListener {

    private val TAG: String = javaClass.simpleName
    private lateinit var prefs: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.act_game)
        if (savedInstanceState == null) {
            ActivityUtils.addFragment(supportFragmentManager, GameFragment(),
                    R.id.fragment_container)
        }
        subscribeUi()
        prefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE)
    }

    override fun onStart() {
        super.onStart()
        // Setup click listeners
        btnBack.setOnClickListener { onBackPressed() }

        if (prefs.getBoolean(PREFS_TUTORIAL, true)) {
            prefs.edit().putBoolean(PREFS_TUTORIAL, false).apply()
            showIntroTutorial()
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        // If it's not a daily puzzle, back to the levels screen
        if (!viewModel.isDailyLevel()) {
            backToLevelsActivity()
        }
    }

    // This method is invoked in the activity when the listener is triggered
    // Access the data result passed to the activity here
    override fun onNextBtnClickedDialog() {
        viewModel.makeLevelSolved()
        if (viewModel.isDailyLevel()) {
            onBackPressed()
        } else {
            // find next level to play
            //Toast.makeText(this, "LOOKING FOR THE NEXT LEVEL", Toast.LENGTH_SHORT).show()
            val nextLevel = viewModel.getNextLevel()
            if (nextLevel != null) {
                startGameActivity(nextLevel.id)
            } else {
                Toast.makeText(
                        this,
                        "Congratulations! You have solved all of the levels. New levels are coming!",
                        Toast.LENGTH_LONG).show()
                onBackPressed()
            }
        }
    }

    override fun onRewardUser() {
        viewModel.increaseHints(1)
        tvHintsCount.text = viewModel.getUser().value?.hints.toString()
    }

    private fun startGameActivity(nextLevelId: String) {
        startActivity(intentFor<GameActivity>(EXTRA_LEVEL_ID to nextLevelId).clearTop())
        finish()
        overridePendingTransition(R.anim.slide_from_right, R.anim.slide_to_left)
    }

    private fun subscribeUi() {
        // Getting level id through the Intent.
        levelId = requireNotNull(intent.getStringExtra(EXTRA_LEVEL_ID),
                { "Activity parameter 'EXTRA_LEVEL_ID' is missing" })

        // Observe updates to the LiveData level.
        viewModel.getLiveLevel().observe(this, Observer<Level> {
            // update UI titlebar with clue
            it?.let { tvTitle.text = it.clue }
        })
        viewModel.getLiveWords().observe(this, Observer {
            it?.let {
                if (viewModel.isLevelSolved()) {
                    showLevelSolvedDialog()
                    viewModel.makeLevelSolved()
                    // restore level
                }
            }
        })
        viewModel.getUser().observe(this, Observer {
            tvHintsCount.text = it?.hints.toString()
        })
    }

    private fun showLevelSolvedDialog() {
        val dialog = LevelSolvedDialog.newInstance(
                Color.parseColor(viewModel.getNextLevelColor()),
                viewModel.isDailyLevel(),
                viewModel.getLiveLevel().value?.state == FINISHED)
        dialog.show(supportFragmentManager, "fragment_level_solved")
        if (viewModel.getLiveLevel().value?.state == IN_PROGRESS) {
            if (viewModel.isDailyLevel()) {
                // adds + 3 HINTS
                viewModel.increaseHints(USER_DAILY_LEVEL_SOLVED_REWARD)
            } else {
                // adds + 2 HINTS
                viewModel.increaseHints(USER_LEVEL_SOLVED_REWARD)
            }
        }
    }

    /**
     * Back navigation. Navigates from GameActivity to LevelsActivity passing Pack id in the Intent.
     */
    private fun backToLevelsActivity() {
        // Passing pack's id through the Intent.
        viewModel.getPackId()?.let {
            startActivity(intentFor<LevelsActivity>(EXTRA_PACK_ID to it).clearTop())
            finish()
            overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right)
        }
    }

    private fun showIntroTutorial() {
        // Show Tutorial
        val fancyShowCaseView1 = FancyShowCaseView.Builder(this)
                .title("In WordChunks, your challenge is to reconstruct hidden words using a grid of chunks.")
                .fitSystemWindows(true)
                .build()

        val fancyShowCaseView2 = FancyShowCaseView.Builder(this)
                .focusOn(tvTitle)
                .focusShape(FocusShape.ROUNDED_RECTANGLE)
                .title("Each level has a specific theme that all words are associated with!")
                .fitSystemWindows(true)
                .build()

        val fancyShowCaseView3 = FancyShowCaseView.Builder(this)
                .focusOn(rvWords)
                .focusShape(FocusShape.ROUNDED_RECTANGLE)
                .title("Each word has the length and optionally the first letter.\nIn order to solve the puzzle, you need to find specific hidden words for each level.")
                .titleStyle(0, Gravity.BOTTOM or Gravity.CENTER)
                .fitSystemWindows(true)
                .build()

        val fancyShowCaseView4 = FancyShowCaseView.Builder(this)
                .focusOn(rvChunks)
                .focusShape(FocusShape.ROUNDED_RECTANGLE)
                .title("You have to use all these chunks.\nCorrect answers accepted automatically.\nThe words order doesn't matter.")
                .titleStyle(0, Gravity.CENTER or Gravity.TOP)
                .fitSystemWindows(true)
                .build()

        val fancyShowCaseView5 = FancyShowCaseView.Builder(this)
                .focusOn(btnHint)
                .focusShape(FocusShape.ROUNDED_RECTANGLE)
                .title("The game will get trickier.\n\nIf you get stuck, you can use hints!")
                .titleStyle(0, Gravity.CENTER or Gravity.TOP)
                .fitSystemWindows(true)
                .build()

        val fancyShowCaseView6 = FancyShowCaseView.Builder(this)
                .focusOn(btnShuffle)
                .focusShape(FocusShape.ROUNDED_RECTANGLE)
                .title("Or use shuffle!\n\nGood luck!")
                .fitSystemWindows(true)
                .build()

        FancyShowCaseQueue()
                .add(fancyShowCaseView1)
                .add(fancyShowCaseView2)
                .add(fancyShowCaseView3)
                .add(fancyShowCaseView4)
                .add(fancyShowCaseView5)
                .add(fancyShowCaseView6)
                .show()
    }
}